### PR TITLE
Added search and banner endpoints. Created KB processing.

### PR DESCRIPTION
**Pull request by Tehsurfer**
# Description

Endpoints added:
 - '/search/<term>' returns a flat dict of results matching <term>
 - '/search/' returns all datasets on scicrunch
 - '/banner/<dataset_id>' return url locations of banners from
   Blacfkfynn

Additional processing added: 
Added a function to flatten nested dictionaries for easier processing in javascript

## Type of change

Delete those that don't apply.

- [ x ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -164,6 +164,40 @@ def sim_dataset(id):
...
+@app.route("/search/<query>")
+def kb_search(query):
+    try:
+        response = requests.get(f'https://scicrunch.org/api/1/elastic/SPARC_Datasets_pr/_search?q={query}&api_key={Config.KNOWLEDGEBASE_KEY}')
+        return process_kb_results_recursive(response.json())
+    except requests.exceptions.HTTPError as err:
+        logging.error(err)
+        return json.dumps({'error': err})
+
+@app.route("/search/")
```
Comment (alan-wu) [inline app/main.py:176]: It is possible to have the query argument here as an optional, that would merge these two methods into one. 
[https://flask.palletsprojects.com/en/master/api/#url-route-registrations](url)
[https://stackoverflow.com/questions/53445145/flask-optional-path-args](url)

File: app/main.py [hunk 2]
```diff
@@ -13,11 +13,11 @@
...
 from app.config import Config
 
-# from blackfynn import Blackfynn
+from blackfynn import Blackfynn
 from app.serializer import ContactRequestSchema
 from scripts.email_sender import EmailSender
-
-# from pymongo import MongoClient
+from app.process_kb_results import process_kb_results_recursive
+from pymongo import MongoClient
```
[the remark this answers is not in the document]
Comment (Tehsurfer) [reply app/main.py:20]: Yeah the response processing is pretty complex so wanted it server side. Thought I would split it into a different file to keep the `app.py` file focused on the routes.

File: app/main.py [hunk 3]
```diff
@@ -80,10 +80,10 @@ def connect_to_blackfynn():
         host=Config.BLACKFYNN_API_HOST
     )
 
-# @app.before_first_request
-# def connect_to_mongodb():
-#     global mongo
-#     mongo = MongoClient(Config.MONGODB_URI)
+@app.before_first_request
+def connect_to_mongodb():
+    global mongo
```
[the remark this answers is not in the document]
Comment (Tehsurfer) [reply app/main.py:85]: Yes it should sorry

File: app/main.py (hunk 1 above)
Comment (Tehsurfer) [reply app/main.py:176]: Oo thanks Alan! I didn't think to try checking if that was possible.

File: app/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alan-wu) [reply app/main.py:85]: Sorry if I misled you in anyway but I want the connect_to_mongdb method to be be commented out, not connect_to_blackfynn.

File: app/main.py [hunk 4]
```diff
@@ -13,11 +13,11 @@
 from blackfynn import Blackfynn
 from app.config import Config
 
-# from blackfynn import Blackfynn
+from blackfynn import Blackfynn
```
Comment (alan-wu) [inline app/main.py:16]: Remove this line please, it is called few lines above.

File: app/main.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alan-wu) [reply app/main.py:20]: Comment out pymongo please, it is not used anywhere.

**Review by alan-wu — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tehsurfer) [reply app/main.py:85]: Ah no that's my fault, I just commented the Blackfynn method out of habit

**Comment by alan-wu** [comment]
Look good, I will merge it in.

**Pull request merged**